clustering_statistics_functions.R

```diff
@@ -0,0 +1,151 @@
+    
+    # true clusters
+    orignal_data <- data.frame(sample_id, metadata_num)
+    
+    # predicted clusters
+    cl_data <- data.frame(sample_id, km$cluster)
+    
+    # calculate NMI and ARI score
+    nmi_score <- NMI::NMI(orignal_data, cl_data)$value
+    ari_score <- mclust::adjustedRandIndex(km$cluster, metadata_num)
+    
+    nmi_score_all <- c(nmi_score_all, nmi_score)
+    ari_score_all <- c(ari_score_all, ari_score)
+  }
+  
+  # Compile all results into a data.frame
+  results <- data.frame(ari = ari_score_all, nmi = nmi_score_all)
+  return(results)
+}
+
+KnnEval <- function(feature, metadata = metadata, iter = 10){
+  # This function performs knn based evaluation 
+  # Args:
+  #  feature: a data.frame contains clustering dimensions
+  #  metadata: vector contains cell type or other metadata information  
+  #  iter: number of interation for knn clustering
+  # Returns:
+  #  list of accuracy scores for each iteration of KNN, as a dataframe
+  
+  # Make the data into a data.frame:
+  feature <- data.frame("tsne" = feature, "metadata" = metadata)
+  
+  # Split observations into groups
+  cv <- cvTools::cvFolds(nrow(feature), K = iter, R = 1)
+  
+  # Create empty objects to store the performance information for each iteration
+  perf.eval <- list()
+  confusion.matrix <- 0
+  
+  # Go through this iteration that k times
+  for (i in 1:iter) {
+    # Isolate samples for training the model
+    train <- feature[cv$subsets[-which(cv$which == i)], ]
+    
+    # Isolate samples for testing the model
+    test <- feature[cv$subsets[which(cv$which == i)], ]
+    
+    # Perform KNN model fitting
+    knn.fit <- caret::train(metadata~. , data = train, method = "knn",
+                            trControl = caret::trainControl(method = "cv", number = 3),
+                            preProcess = c("center", "scale"),
+                            tuneLength = 10)
+    
+    # Evaluate the model
+    knn.pred <- predict(knn.fit, newdata = subset(test, select = -c(metadata)))
+    perf.eval[[i]] <- round(cal_performance(knn.pred, test$metadata, 3), 2)
+    
+    # Make the results into a matrix
+    matrix <- as.matrix(table(test$metadata, knn.pred, deparse.level = 0))
+    confusion.matrix <- matrix + confusion.matrix
+  }
+  
+  # Get mean performance of cross validation
+  perf.eval <- dplyr::bind_rows(perf.eval)
+  accuracy <- perf.eval$accuracy
+  
+  return(data.frame("knn" = accuracy))
+}
+
+
+PlotResults <- function(results.list, name = "results") {
+  # This function makes a boxplot of the cluster statistic results 
+  # Args:
+  #  results.list: a list of dataframes which each column contains a different 
+  #                statistic for 10 iterations
+  #  name: name to use for the png to be saved and the plot title 
+  # Returns: boxplots of the normalization method cluster statistics. Prints the
+  #          plots and save the plot as a png
+  
+  # Get meta info
+  meta <- names(unlist(results.list))
```

Comment:
Because I didn't use `dplyr::melt.list` instead and use this to retrieve the names of all the data. This isn't the most elegant way to do this. I will go back and make this less "janky"